untyped-lambda-calculus/article.md

```diff
@@ -0,0 +1,493 @@
+# Untyped Lambda Calculus
+
+When talking about Haskell, the term "lambda calculus" often crops up.
+It is a theoretical framework used to define the meaning of "computation" in many functional languages.
+Understanding lambda calculus can be very helpful when talking about Haskell programs.
+It is also used extensively in programming language theory research and functional language design.
+
+There are two broad kinds of lambda calculus: _untyped_ and _typed_.
+The _untyped_ kind, discussed in this article, expresses an unrestricted computation.
+However, as a system of logic, it is self-contradictory.
+The _typed_ kind, which we intend to discuss in a follow-up article, is a restricted version.
+It doesn't have the logical paradox that plagues the untyped kind and can serve as a logic system.
+
+The list of languages based on lambda calculus includes:
+
+- Haskell,
+- Most languages in the ML family, like OCaml and F#,
+- Most dependently typed languages, like Agda, Coq, F*, Idris,
+- Cardano blockchain's Plutus Core.
+
+There's a rumor that John McCarthy based Lisp on lambda calculus, but this seems to be, broadly speaking, false.
+The original Lisp used different name resolution semantics, and as much was stated by John himself:
+
+> ... one of the myths concerning LISP that people think up or invent for themselves becomes apparent, and that is that LISP is somehow a realization of the lambda calculus, or that was the intention.
+> The truth is that I didn't understand the lambda calculus, really.
+>
+> -- John McCarthy, Lisp session, History of Programming Languages
+
+Also, while not strictly speaking based on lambda calculus, C++ template metalanguage is a functional language in its own right, so much of what we discuss below happens to be applicable.
```

Comment:
Is this necessary? I feel like we can cut this.

Reply:
While not strictly speaking _necessary_ (nothing in the intro is strictly speaking necessary besides "λ-calculus is every FP programming language invented, ever" :laughing: ), I, for one, find this curious, and it suggests that lambda-calculus is applicable outside the hard-core FP bubble. If the point of the intro is to provide the motivation for the reader, I think this paragraph is useful. I'd rather trim down the LISP discussion if you think the intro is a bit too long.

Reply:
In that case, perhaps we can just follow this comment: https://github.com/serokell/blog-posts/pull/12#discussion_r857395866